packages/service-library/src/servicelib/utils.py

```diff
@@ -175,3 +176,115 @@ def unused_port() -> int:
     with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
         s.bind(("127.0.0.1", 0))
         return cast(int, s.getsockname()[1])
+
+
+T = TypeVar("T")
+
+
+async def limited_as_completed(
+    awaitables: Iterable[Awaitable[T]] | AsyncIterable[Awaitable[T]], *, limit: int = 1
+) -> AsyncGenerator[asyncio.Future[T], None]:
+    """Runs awaitables using limited concurrent tasks and returns
+    result futures unordered.
+
+    Arguments:
+        awaitables -- The awaitables to limit the concurrency of.
+
+    Keyword Arguments:
+        limit -- The maximum number of awaitables to run concurrently.
+                0 or negative values disables the limit. (default: {1})
+
+    Returns:
+        nothing
+
+    Yields:
+        Future[T]: the future of the awaitables as they appear.
+
+
+    """
+    try:
+        awaitable_iterator = aiter(awaitables)  # type: ignore[arg-type]
+        is_async = True
+    except TypeError:
+        assert isinstance(awaitables, Iterable)  # nosec
+        awaitable_iterator = iter(awaitables)  # type: ignore[assignment]
+        is_async = False
+
+    completed_all_awaitables = False
+    pending_futures: set[asyncio.Future] = set()
+
+    while pending_futures or not completed_all_awaitables:
+        while (
+            limit < 1 or len(pending_futures) < limit
+        ) and not completed_all_awaitables:
+            try:
+                aw = (
+                    await anext(awaitable_iterator)
+                    if is_async
+                    else next(awaitable_iterator)  # type: ignore[call-overload]
```

Comment:
I did not get why we have `AsyncIterable` and `Iterable` objects here. The function signature clearly hints that it should work with `AsyncIterable` only. It's slightly confusing.

Reply:
well handling of async iterable and non-async iterable is different. one use next and the other anext.